include/exec/system_context.hpp

```diff
@@ -83,19 +100,118 @@ namespace exec {
     template <class _Sender>
     using __sender_data_t = decltype(stdexec::sync_wait(std::declval<_Sender>()).value());
 
+    /// A very simple spin lock. It is used in leu of a mutex because
+    /// instances of this class are constructed and never destroyed.
+    /// Doing that with a std::mutex might(?) leak system resources.
+    struct __spin_lock {
+      constexpr __spin_lock() noexcept = default;
+
+      void __lock() noexcept {
+        bool __is_locked = false;
+        // spin until we get the lock:
+        while (!__locked_.compare_exchange_strong(__is_locked, true)) {
+          std::this_thread::yield();
+          __is_locked = false;
+        }
+      }
+
+      void __unlock() noexcept {
+        __locked_ = false;
+      }
+
+     private:
+      std::atomic<bool> __locked_{false};
+    };
   } // namespace __detail
 
   class system_scheduler;
   class system_sender;
-  template <stdexec::sender _S, std::integral _Size, class _Fn>
+  template <stdexec::sender _Sndr, std::integral _Size, class _Fn>
   class system_bulk_sender;
 
+  /// @brief A base class that helps populate the system_context_interface
+  ///        using members of the derived class.
+  class system_context_base : public system_context_interface {
```

Comment:
I don't think this type should be part of the public interface.
Can we name it with underscores instead? i.e. `__system_context_base`

Reply:
no, it is intended to be a public part of the API. it is a utility to ensure that the "virtual table" is filled out correctly, the version number is set appropriately, and that the ref count is properly set to 1.

Reply:
So are users implementing their own system-context _required_ to use this?
Or are we defining the semantics enough that it is well-specified how to initialize a `system_context_interface` structure directly?

Reply:
> So are users implementing their own system-context required to use this?

no